src/common/src/types/to_sql.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::error::Error;
+
+use bytes::BytesMut;
+use postgres_types::{to_sql_checked, IsNull, ToSql, Type};
+
+use crate::types::ScalarRefImpl;
+
+impl ToSql for ScalarRefImpl<'_> {
+    to_sql_checked!();
+
+    fn to_sql(&self, ty: &Type, out: &mut BytesMut) -> Result<IsNull, Box<dyn Error + Sync + Send>>
+    where
+        Self: Sized,
+    {
+        match self {
+            ScalarRefImpl::Int16(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Int32(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Int64(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Serial(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Float32(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Float64(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Utf8(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Bool(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Decimal(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Interval(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Date(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Timestamp(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Timestamptz(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Time(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Bytea(v) => v.to_sql(ty, out),
+            ScalarRefImpl::Jsonb(_) // jsonbb::Value doesn't implement ToSql yet
+            | ScalarRefImpl::Int256(_)
+            | ScalarRefImpl::Struct(_)
+            | ScalarRefImpl::List(_) => {
+                bail_not_implemented!("the postgres encoding for {ty} is unsupported")
+            }
+        }
+    }
+
+    // return true to accept all types
+    fn accepts(_ty: &Type) -> bool
+    where
+        Self: Sized,
+    {
+        true
```

Comment:
This makes me think we probably should not `impl ToSql for ScalarRefImpl`.

It looks like `tokio-postgres` actually uses `&dyn ToSql` (`BorrowToSql`). That is, we may be able to `impl BorrowToSql for DatumRef` without having `impl ToSql for ScalarRefImpl`. `ToSql` is not supposed to be defined for a weak type. I will give it a try and let you know if it is doable.

Reply:
A stronger `accepts` test under the `&dyn ToSql` signature is possible for `ScalarImplRef` but not `DatumRef`, because we have not type info when it is `null`. In the latter case we have to accept all types and return `true` blindly.

It is possible to pass the type info alongside `start_pk_row` all the way down from `CdcBackfillExecutor` when doing snapshot read, as executors always know the data types of all columns. But we can defer that refactor until data type info is really necessary. For now, let just leave a brief comment:
> `DatumRef::None` has to accept all types when associated `DataType` is not available.